L1Trigger/Phase2L1GT/python/l1tGTAlgoBlockProducer_cff.py

```diff
@@ -0,0 +1,23 @@
+import FWCore.ParameterSet.Config as cms
+import re
+
+algorithms = cms.VPSet()
+
+l1tGTAlgoBlockProducer = cms.EDProducer(
+    "L1GTAlgoBlockProducer",
+    algorithms = algorithms
+)
+
+def collectAlgorithmPaths(process) -> "tuple[cms.Path]":
+    str_paths = set()
+    for algorithm in algorithms:
+        algo_paths = re.sub(r'[()]'," " , algorithm.expression.value()).split()
+        for algo in algo_paths:
+            if algo in process.pathNames() :
+                str_paths.add(algo)
+    paths = set()
+
+    for str_path in str_paths:
+        paths.add(getattr(process, str_path))
+
+    return tuple(paths)
```

Comment:
Unfortunately this approach can potentially return the `paths` in a different order on each invocation.
This does happen in production, and results in jobs with a different process hash.
The framework then considers them different configurations and does not fully merge the resulting files; for example, they appear to belong to different runs (even though the run number is the same).

Reply:
@fwyzard I see, so in principle if the order is deterministic there shouldn't be a problem, right? So wouldn't a simple `sorted(str_paths)` then do the trick?

Reply:
Yes, that should fix the reproducibility problem.

Reply:
@fwyzard Would you mind either checking whether this indeed solves the issue or pointing me to how I can reproduce it myself before submitting a PR?

Reply:
I'm not sure.

You could try running the `cmsDriver` command used in production a few times, potentially on different machines (like `lxplus8` and `lxplus9`) and check
  - that sometimes you get different ordering with the current code
  - that you always get the same order with the fix

Reply:
@rovere found that the ordering of the L1 paths in the Phase 2 production workflows is not consistent, and pointed me to this snippet.

Reply:
Depending on how it got run, I'm the one to blame instead: https://github.com/cms-l1t-offline/cmssw/blob/315e0b6cbcc55bec5b5eecbe955f25f68db9daef/Configuration/Applications/python/ConfigBuilder.py#L1523-L1579

Reply:
I guess a 
```python
triggerScheduleList = [getattr(self.process, name) for name in sorted(triggerPaths)]
```
would also do the trick here.

Reply:
@rovere Can you confirm whether the skims you were making used `cmsDriver` instead of putting all this together by hand using the GT team function (seems unlikely to me).

We could sort the trigger paths I guess, which is deterministic, I assume this order doesn't matter in the grand scheme of emulating L1T/P2GT?

Reply:
Nope

> @rovere Can you confirm whether the skims you were making used `cmsDriver` instead of putting all this together by hand using the GT team function (seems unlikely to me).
> 
> We could sort the trigger paths I guess, which is deterministic, I assume this order doesn't matter in the grand scheme of emulating L1T/P2GT?

Nope, the order doesn't matter.

Reply:
> @rovere found that the ordering of the L1 paths in the Phase 2 production workflows is not consistent, and pointed me to this snippet.

Does this occur between workflows, or already between jobs within one workflow?

Reply:
From what I understood, this happens in different jobs of the "step2" production workflows.

Reply:
Hmh, I though production (or RelVal) would run `cmsDriver` and pickling, and thus the randomization of the order, only once for one workflow, and all jobs of the workflow would use the same pickled configuration. Oh well.

In any case jobs of one workflow having randomized order would be Really Bad (for those interested in a a previous episode in this topic see https://github.com/cms-sw/cmssw/pull/42833).

Reply:
It is already happening in production with the Spring2024 Phase-2 samples:
```
$ for F in /eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAO
D/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/*.root; do echo $F; edmProvDump $F | grep 'Processing History:' -A3; echo; done
/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/0071b62d-2098-425d-bf40-b8ee556a4bdf.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (0f408e19d0a839b8572a8631c9f8cdc9)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/0101febf-d45c-4198-ac2a-67293cf30a6a.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (900885b53249a402aea6fb2a33b4e332)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/011502f3-8642-4d14-9ac8-222b10871256.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (5a5f82bf8e97542e86c5a69f6a863628)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/0132180b-40d0-40f3-8fde-3b6d11b352ab.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (91d750b7abc8bcfa98cc363ace5a7ec4)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/015898ed-d90c-4a6f-b1cd-b60addb3a44d.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (ef5538925a0cf86e10c192e7e5f28796)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/01c1c0f3-f22b-4d5b-a410-636acb0f977d.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (6f7ec9bf1640330e4d4535b2e0b92a90)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/01e320cc-b191-4d21-bfce-cb03a0ac0709.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (0f408e19d0a839b8572a8631c9f8cdc9)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

/eos/cms/store/mc/Phase2Spring24DIGIRECOMiniAOD/MinBias_TuneCP5_14TeV-pythia8/GEN-SIM-DIGI-RAW-MINIAOD/PU200ALCA_140X_mcRun4_realistic_v4-v2/120007/0285ba92-3483-48c6-804a-40e8989ab5aa.root
Processing History:
  SIM '' '"CMSSW_14_0_9"' [1]  (3e9adf1abe9d86a46e0052e939922667)
    HLT '' '"CMSSW_14_0_9"' [1]  (b20c521d9913c180494dccd55a6032e8)
      RECO '' '"CMSSW_14_0_9"' [1]  (85f398480a69ab0c4368f75496dcd008)

...
```

The `HLT` step has a different hash in each file, at least in the first eight, then I stopped checking.

Dumping and comparing the ParameterSet associated to the different hashes, the differences are in the order of the paths:
```diff
   @all_sources: vstring tracked  = {'@main_input'}
   @end_paths: vstring tracked  = {'HLTAnalyzerEndpath','endjob_step'}
-  @paths: vstring tracked  = {'digitisation_step','L1TrackTrigger_step','L1simulation_step','digi2raw_step','GTemulation_step','pDoubleTkMuon15_7','pQuadJet70_55_40_40','pSingleTkMuon22','pPuppiHT450','pSingleIsoTkPho36','pSinglePuppiJet230','pSingleEGEle51','pTripleTkMuon5_3_3','pDoubleEGEle37_24','pDoubleTkEle25_12','pSingleIsoTkEle28','pPuppiHT400','pDoubleIsoTkPho22_12','pSingleTkEle36','pIsoTkEleEGEle22_12','pDoublePuppiTau52_52','pPuppiMET200','L1T_SinglePFPuppiJet230off','L1T_PFPuppiHT450off','L1T_PFPuppiMET220off','HLT_AK4PFPuppiJet520','HLT_PFPuppiHT1070','HLT_PFPuppiMETTypeOne140_PFPuppiMHT140','L1T_PFHT400PT30_QuadPFPuppiJet_70_55_40_40_2p4','L1T_DoublePFPuppiJets112_2p4_DEta1p6','HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepCSV_2p4','HLT_PFHT200PT30_QuadPFPuppiJet_70_40_30_30_TriplePFPuppiBTagDeepCSV_2p4','HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepCSV_2p4','HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepFlavour_2p4','HLT_PFHT200PT30_QuadPFPuppiJet_70_40_30_30_TriplePFPuppiBTagDeepFlavour_2p4','HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepFlavour_2p4','L1T_SingleTkMuon_22','L1T_DoubleTkMuon_15_7','L1T_TripleTkMuon_5_3_3','HLT_Mu50_FromL1TkMuon','HLT_IsoMu24_FromL1TkMuon','HLT_Mu37_Mu27_FromL1TkMuon','HLT_Mu17_TrkIsoVVL_Mu8_TrkIsoVVL_DZ_FromL1TkMuon','HLT_TriMu_10_5_5_DZ_FromL1TkMuon','L1T_TkEm51','L1T_TkEle36','L1T_TkIsoEm36','L1T_TkIsoEle28','L1T_TkEm37TkEm24','L1T_TkEle25TkEle12','L1T_TkIsoEm22TkIsoEm12','L1T_TkIsoEle22TkEm12','HLT_Ele32_WPTight_Unseeded','HLT_Ele26_WP70_Unseeded','HLT_Photon108EB_TightID_TightIso_Unseeded','HLT_Photon187_Unseeded','HLT_DoubleEle25_CaloIdL_PMS2_Unseeded','HLT_Diphoton30_23_IsoCaloId_Unseeded','HLT_Ele32_WPTight_L1Seeded','HLT_Ele115_NonIso_L1Seeded','HLT_Ele26_WP70_L1Seeded','HLT_Photon108EB_TightID_TightIso_L1Seeded','HLT_Photon187_L1Seeded','HLT_DoubleEle25_CaloIdL_PMS2_L1Seeded','HLT_DoubleEle23_12_Iso_L1Seeded','HLT_Diphoton30_23_IsoCaloId_L1Seeded','HLT_DoubleMediumChargedIsoPFTauHPS40_eta2p1','HLT_DoubleMediumDeepTauPFTauHPS35_eta2p1','MC_JME','MC_BTV','MC_Ele5_Open_Unseeded','MC_Ele5_Open_L1Seeded','HLTriggerFinalPath','HLTAnalyzerEndpath','endjob_step'}
+  @paths: vstring tracked  = {'digitisation_step','L1TrackTrigger_step','L1simulation_step','digi2raw_step','GTemulation_step','pPuppiHT450','pSingleIsoTkPho36','pSingleEGEle51','pSinglePuppiJet230','pTripleTkMuon5_3_3','pDoubleEGEle37_24','pSingleIsoTkEle28','pIsoTkEleEGEle22_12','pDoubleTkEle25_12','pSingleTkEle36','pPuppiHT400','pDoubleIsoTkPho22_12','pQuadJet70_55_40_40','pPuppiMET200','pDoublePuppiTau52_52','pDoubleTkMuon15_7','pSingleTkMuon22','L1T_SinglePFPuppiJet230off','L1T_PFPuppiHT450off','L1T_PFPuppiMET220off','HLT_AK4PFPuppiJet520','HLT_PFPuppiHT1070','HLT_PFPuppiMETTypeOne140_PFPuppiMHT140','L1T_PFHT400PT30_QuadPFPuppiJet_70_55_40_40_2p4','L1T_DoublePFPuppiJets112_2p4_DEta1p6','HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepCSV_2p4','HLT_PFHT200PT30_QuadPFPuppiJet_70_40_30_30_TriplePFPuppiBTagDeepCSV_2p4','HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepCSV_2p4','HLT_PFHT330PT30_QuadPFPuppiJet_75_60_45_40_TriplePFPuppiBTagDeepFlavour_2p4','HLT_PFHT200PT30_QuadPFPuppiJet_70_40_30_30_TriplePFPuppiBTagDeepFlavour_2p4','HLT_DoublePFPuppiJets128_DoublePFPuppiBTagDeepFlavour_2p4','L1T_SingleTkMuon_22','L1T_DoubleTkMuon_15_7','L1T_TripleTkMuon_5_3_3','HLT_Mu50_FromL1TkMuon','HLT_IsoMu24_FromL1TkMuon','HLT_Mu37_Mu27_FromL1TkMuon','HLT_Mu17_TrkIsoVVL_Mu8_TrkIsoVVL_DZ_FromL1TkMuon','HLT_TriMu_10_5_5_DZ_FromL1TkMuon','L1T_TkEm51','L1T_TkEle36','L1T_TkIsoEm36','L1T_TkIsoEle28','L1T_TkEm37TkEm24','L1T_TkEle25TkEle12','L1T_TkIsoEm22TkIsoEm12','L1T_TkIsoEle22TkEm12','HLT_Ele32_WPTight_Unseeded','HLT_Ele26_WP70_Unseeded','HLT_Photon108EB_TightID_TightIso_Unseeded','HLT_Photon187_Unseeded','HLT_DoubleEle25_CaloIdL_PMS2_Unseeded','HLT_Diphoton30_23_IsoCaloId_Unseeded','HLT_Ele32_WPTight_L1Seeded','HLT_Ele115_NonIso_L1Seeded','HLT_Ele26_WP70_L1Seeded','HLT_Photon108EB_TightID_TightIso_L1Seeded','HLT_Photon187_L1Seeded','HLT_DoubleEle25_CaloIdL_PMS2_L1Seeded','HLT_DoubleEle23_12_Iso_L1Seeded','HLT_Diphoton30_23_IsoCaloId_L1Seeded','HLT_DoubleMediumChargedIsoPFTauHPS40_eta2p1','HLT_DoubleMediumDeepTauPFTauHPS35_eta2p1','MC_JME','MC_BTV','MC_Ele5_Open_Unseeded','MC_Ele5_Open_L1Seeded','HLTriggerFinalPath','HLTAnalyzerEndpath','endjob_step'}
   @process_name: string tracked  = 'HLT'
   GTemulation_step: vstring tracked  = {}
```

or maybe more readable
```diff
--- 0f408e19d0a839b8572a8631c9f8cdc9    2024-10-07 19:42:42.000000001 +0200
+++ 5a5f82bf8e97542e86c5a69f6a863628    2024-10-07 19:42:53.000000001 +0200
@@ -4,23 +4,23 @@
     'L1simulation_step',
     'digi2raw_step',
     'GTemulation_step',
-    'pDoubleTkMuon15_7',
-    'pQuadJet70_55_40_40',
-    'pSingleTkMuon22',
     'pPuppiHT450',
     'pSingleIsoTkPho36',
-    'pSinglePuppiJet230',
     'pSingleEGEle51',
+    'pSinglePuppiJet230',
     'pTripleTkMuon5_3_3',
     'pDoubleEGEle37_24',
-    'pDoubleTkEle25_12',
     'pSingleIsoTkEle28',
+    'pIsoTkEleEGEle22_12',
+    'pDoubleTkEle25_12',
+    'pSingleTkEle36',
     'pPuppiHT400',
     'pDoubleIsoTkPho22_12',
-    'pSingleTkEle36',
-    'pIsoTkEleEGEle22_12',
-    'pDoublePuppiTau52_52',
+    'pQuadJet70_55_40_40',
     'pPuppiMET200',
+    'pDoublePuppiTau52_52',
+    'pDoubleTkMuon15_7',
+    'pSingleTkMuon22',
     'L1T_SinglePFPuppiJet230off',
     'L1T_PFPuppiHT450off',
     'L1T_PFPuppiMET220off',
```

Reply:
Thanks for the details!

Reply:
@fwyzard @makortel would it be appropriate then to just open a quick PR for @HaarigerHarald 's solution:

```python
triggerScheduleList = [getattr(self.process, name) for name in sorted(triggerPaths)]
```

Just guarantee that these are always added to the schedule from the same alphabetically sorted list? I can knock that out quickly if desired.

Reply:
If from the L1T side people are not interested in a specific order (other than alphabetical), then sure, please go ahead.

Reply:
> If from the L1T side people are not interested in a specific order (other than alphabetical), then sure, please go ahead.

For us the "true" order of algorithms is how they are stored in the `P2GTAlgoBlockMap` and this order doesn't depend on the schedule order or path order.

Reply:
Would it be feasible to insert them in the `process` using the same order ?
Or it's not meaningful anyway ?

Reply:
There is a I suppose it could be ordered by seed number in the menu, but I don't think that is meaningful enough in this context that we truly _need_ to order them that way as paths as well.

Reply:
> Would it be feasible to insert them in the `process` using the same order ? Or it's not meaningful anyway ?

Given that an algorithm can be a combination of multiple paths and in principle a path can be used in multiple algorithms (although it's unclear whether we should deem the latter case bad practice), I'm not entirely sure how meaningful any path order actually is.

Reply:
I have opened https://github.com/cms-sw/cmssw/pull/46303. @fwyzard or @rovere would it be possible to request that one of you retests the issues with this PR pulled in? I'm not quite sure what you have done to recreate this issue.

Reply:
@rovere simply used the samples from the official Spring24 production; a test would be to re-produce the datasets.